### PR TITLE
webgpu: support MaxPoolGrad kernel

### PR DESCRIPTION
**Pull request by xhcao**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/7261)

### CODE REVIEW

File: tfjs-backend-webgpu/src/kernels/MaxPoolGrad.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    {type: 'int32', data: [convInfo.dilationHeight, convInfo.dilationWidth]},
+    {type: 'int32', data: [convInfo.inHeight, convInfo.inWidth]}, {
+      type: 'int32',
+      data: [convInfo.effectiveFilterHeight, convInfo.effectiveFilterWidth]
+    }
+  ];
+  const maxPoolPositions = backend.runWebGPUProgram(
+      maxPoolPositionsProgram, [x], 'int32', uniformData);
+
+  const avgPoolBackpropProgram = new MaxPool2DBackpropProgram(convInfo);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/MaxPoolGrad.ts:53]: nit: `avgPoolXXX` -> `maxPoolXXX`

File: tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/**
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts:1]: Will `MaxPool3DBackpropProgram` share the same file? If yes, maybe we should remove `2d` from file name?

File: tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+
+export class MaxPool2DBackpropProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['dy', 'maxPos'];
+  uniforms =
+      `stride : vec2<i32>, pad : vec2<i32>, dilation : vec2<i32>, filterDims : vec2<i32>,
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts:29]: Can we unify to use `strides`, `pads`, `dilations` from this PR? And you can do the refactor for other places  in a separate PR.

File: tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts (hunk 2 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts:1]: Currently, I am not sure the file will be shared by `3d`, so we keep the name with `2d` here. If we put the `3d` in this file, then we could rename the file.

**Review by qjia7 — APPROVED** [review]

**Comment by xhcao** [comment]
@gyagp Please take a look, thanks

File: tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  constructor(convInfo: backend_util.Conv2DInfo) {
+    this.outputShape = convInfo.inShape;
+
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+
+    this.shaderKey = `max_pool2d_backprop`;
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/max_pool2d_backprop_webgpu.ts:42]:
```suggestion
    this.shaderKey = `maxPool2DBackprop`;
```
What's the convention for shaderKey? Maybe className.replace('Program', '') and variants split by '_'? We need to somehow unify this.

**Review by gyagp — APPROVED** [review]
LGTM with nit to unify shaderKey

**Pull request merged**